File: server/src/main/java/org/jetbrains/bsp/bazel/server/bloop/ScalaInterop.java
package org.jetbrains.bsp.bazel.server.bloop;

import com.google.common.collect.Lists;
import io.vavr.control.Option;
import java.nio.file.Path;
import java.util.Arrays;
import java.util.Collections;
import java.util.Iterator;
import scala.collection.immutable.List$;
import scala.jdk.javaapi.CollectionConverters;

class ScalaInterop {
  private ScalaInterop() {}

  public static <T> scala.collection.immutable.List<T> emptyList() {
    return List$.MODULE$.empty();
  }

  public static <T> scala.Option<T> toOption(Option<T> opt) {
    return opt.fold(scala.Option::empty, scala.Some::apply);
  }

  public static scala.collection.immutable.List<Path> toList(Path input) {
    return toList(Collections.singletonList(input));
  }

  public static <T> scala.collection.immutable.List<T> toList(Iterable<T> input) {
    return CollectionConverters.asScala(input).toList();
  }

  public static <T> scala.collection.immutable.List<T> toList(Iterator<T> input) {
    return CollectionConverters.asScala(input).toList();
  }

  public static <T> scala.collection.immutable.List<T> toList(T first, T... rest) {
    var lst = Lists.<T>newArrayList();
    lst.add(first);
    lst.addAll(Arrays.asList(rest));
    return toList(lst);
  }
}

File: server/src/main/java/org/jetbrains/bsp/bazel/server/bloop/ScalaInterop.kt
package org.jetbrains.bsp.bazel.server.bloop

import scala.Option
import scala.Some
import scala.collection.immutable.List
import scala.collection.immutable.`List$`
import scala.jdk.javaapi.CollectionConverters.asScala
import java.nio.file.Path

object ScalaInterop {
    fun <T> emptyList(): List<T> = `List$`.`MODULE$`.empty()

    fun <T> emptyOption(): Option<T> = Option.empty()

    fun <T> T?.toOption(): Option<T> =
        this?.let { Some.apply(it) } ?: Option.empty()

    fun Path.toScalaList(): List<Path> =
        listOf(this).toScalaList()

    fun <T> Iterable<T>.toScalaList(): List<T> =
        asScala(this).toList()

    fun <T> Iterable<T>.toListOption(): Option<List<T>> =
        asScala(this).toList().toOption()

    fun <T> Iterator<T>.toScalaList(): List<T> =
        asScala(this).toList()

    fun <T> scalaListOf(vararg rest: T): List<T> =
        rest.toList().toScalaList()
}
